### ISSUE TITLE
Trying to get whole program IR for klee-uclibc

### ISSUE THREAD

**Issue by DaryllRicalde** [travitch/whole-program-llvm#148]
I'm trying to get the whole program IR for [klee-uclibc](https://github.com/klee/klee-uclibc) using wllvm in a similar way to [musllvm ](https://github.com/SRI-CSL/musllvm). I've been able to produce a bitcode archive but am met with an error when trying to run `extract-bc`:

```bash
CC=wllvm WLLVM_CONFIGURE_ONLY=1 ./configure --make-llvm-lib --with-llvm-config /lib/llvm-14/bin/llvm-config --with-cc clang-14
# libc.a is then stored in lib/
extract-bc lib/libc.a
objdump: locale_data.os: file format not recognised
ERROR: Could not dump b'locale_data.os'
```

**Comment by ianamason** [comment]
What is the file type of locale_data.os?

**Comment by DaryllRicalde** [comment]
```bash
$ file klee-uclibc/extra/locale/locale_data.os
locale_data.os: LLVM IR bitcode
```
I was also able to `llvm-dis` it

**Comment by ianamason** [comment]
I'll take a look soonish, but my guess is that it thinks it's a object file, and is trying to extract the paths to the constituent bitcode from the wllvm segment. It might help to have a more detailed log. Can you turn on debugging and get one?

**Comment by DaryllRicalde** [comment]
Thanks, will do.

**Comment by DaryllRicalde** [comment]
Hi, kindly see the (hopefully) more detailed log. 
```bash
DEBUG::extraction.process_file_unix() at extraction.py:669 ::Detected file type is ARCHIVE                                         
objdump: locale_data.os: file format not recognized                                                                                
ERROR::extraction.getSectionSizeAndOffset() at extraction.py:72 ::Could not dump b'locale_data.os'
DEBUG::extraction.handleArchiveLinux() at extraction.py:510 ::Deleting temporary folder "/tmp/tmpiyhiiu6iwllvm" 
```

**Comment by ianamason** [comment]
In the meantime, you might try gllvm. We spend more time on keeping that guy working.

**Comment by DaryllRicalde** [comment]
Cheers Ian. While we're on the topic, have you guys been successful (or attempted) in building a whole program IR for other libc implementations apart from musl?

**Comment by ianamason** [comment]
Not sure anyone here is still doing this stuff. What is the libc in alpine linux?

**Comment by DaryllRicalde** [comment]
I actually don't know for certain but a quick Google search says its musl.

**Comment by DaryllRicalde** [comment]
Hi @ianamason I also tried with `gllvm` but was unsuccessful. See the following debug logs

[After running configure with `CC=gclang CXX=gclang++ WLLVM_CONFIGURE_ONLY=1`](https://github.com/travitch/whole-program-llvm/files/12782200/out.txt)

[After running`get-bc lib/libc.a`](https://github.com/travitch/whole-program-llvm/files/12782204/get-bc.txt)

**Comment by ianamason** [comment]
Those STRIP in the build look a little worrying.

The get-bc log is empty?

**Comment by DaryllRicalde** [comment]
Oops sorry about that. 

tl;dr: A lot of `ERROR:ELF file *.os could not be read.` 
Ends with
```bash
artifactFiles:
[]
bcFiles:
[]
ERROR:No bitcode files found
``` 

[get-bc.txt](https://github.com/travitch/whole-program-llvm/files/12782529/get-bc.txt)

**Comment by ianamason** [comment]
I suppose the easiest way for me to see what is going on would be to do it myself.
What is your setup? Where do I get the thing you are compiling etc...

**Comment by DaryllRicalde** [comment]
I'm running Windows Subsystem for Linux 1 (WSL 1)
- Ubuntu 22.04.2 LTS
```bash
$ git clone https://github.com/klee/klee-uclibc.git
cd klee-uclibc
```
Copy these steps https://github.com/travitch/whole-program-llvm/issues/148#issue-1917736325 (or use gllvm)

Let me know if you need further info

**Comment by ianamason** [comment]
It doesn't configure on my M1 laptop, but if I were you I would be building the native version, I have no idea what 
``
--make-llvm-lib
``
does, but these wrapper tools expect to be producing object files. I'm sure there are tutorials out there building uclibc for klee. That use to be a very common thing to do. Good luck.

**Comment by ianamason** [comment]
But my guess is getting the bitcode for this build is probably easier than you are expecting.
You should look into the klee documentation a bit.
